[Liquid Glass] [iOS] Scroll pockets disappear after ending a navigation swipe

whsieh opened the pull request:
#### 0a7fbd0e87756615ec98e0f674be8886bbac17ab
<pre>
[Liquid Glass] [iOS] Scroll pockets disappear after ending a navigation swipe
<a href="https://bugs.webkit.org/show_bug.cgi?id=296254">https://bugs.webkit.org/show_bug.cgi?id=296254</a>
<a href="https://rdar.apple.com/156274449">rdar://156274449</a>

Reviewed by Abrar Rahman Protyasha.

Work around another case where scroll pockets disappear on iOS, when swiping backwards/forwards.
After recent changes in UIKit, pocketed element bounds are cached per scroll pocket, but these
bounds are (sometimes) not invalidated when scroll views change positions/transforms or are
reparented in the view hierarchy. One of these cases is when swiping back/forwards, where
`WKScrollView` is temporarily parented under the `LiveSwipeViewClipping` layer before being moved
back under the `WKWebView` after the swipe is complete. The end result is that scroll pockets are
*removed* from the view hierarchy, since UIKit determines they aren&apos;t necessary, due to the fact
that the pocketed element rects don&apos;t intersect with the scroll view&apos;s frame right after the swipe
finishes, since the scroll view is offscreen.

As a temporary workaround, we forcibly invalidate the scroll pockets&apos; cached geometries after a
navigation swipe finishes. This change should be reverted once the fix for the underlying issue
lands, in <a href="https://rdar.apple.com/156271879">rdar://156271879</a>.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _navigationGestureDidEnd]):
(-[WKWebView _forceScrollPocketsToRecomputeElementRegions]):
* Source/WebKit/UIProcess/ios/UIKitUtilities.h:

Canonical link: <a href="https://commits.webkit.org/297668@main">https://commits.webkit.org/297668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39f6ce8d1699ac37b29197776b961e33dcb8ccd0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112481 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32213 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22690 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118640 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62927 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32865 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40776 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/85619 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36219 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115428 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101200 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65898 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25533 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62423 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95625 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121901 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29463 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94486 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39937 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94222 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24052 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39334 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17132 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35624 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39443 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44931 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/135327 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39080 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/135327 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42415 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40821 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3547,8 +3547,45 @@ - (void)_navigationGestureDidEnd
+        UIRectEdgeLeft,
+        UIRectEdgeBottom,
+        UIRectEdgeRight
+    };
```
I wonder if this could live in UIKitUtilities. I don't know if we use it frequently enough though.

whsieh replied:
Seems reasonable to me — will move it there.

Thanks for the review!

aprotyas commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3547,8 +3547,45 @@ - (void)_navigationGestureDidEnd
+            continue;
+
+        if (![scrollPocket respondsToSelector:@selector(invalidateAllElements)])
+            return;
```
nit: if all scroll pockets are the same impl, maybe put this at the start of this method instead in the for loop?

```objc
if ([_UIScrollPocket instancesRespondToSelector:@selector(invalidateAllElements)])
```

whsieh replied:
Oh, I opted not to go for this approach to avoid having to soft-link `_UIScrollPocket` (since `+instancesRespondToSelector` would be a class method). It's true that they are the same implementation though — I'll instead cache this in a `static bool` so that we avoid checking `-respondsToSelector:` every time we loop here.

whsieh committed:
[Liquid Glass] [iOS] Scroll pockets disappear after ending a navigation swipe
https://bugs.webkit.org/show_bug.cgi?id=296254
rdar://156274449

Reviewed by Abrar Rahman Protyasha.

Work around another case where scroll pockets disappear on iOS, when swiping backwards/forwards.
After recent changes in UIKit, pocketed element bounds are cached per scroll pocket, but these
bounds are (sometimes) not invalidated when scroll views change positions/transforms or are
reparented in the view hierarchy. One of these cases is when swiping back/forwards, where
`WKScrollView` is temporarily parented under the `LiveSwipeViewClipping` layer before being moved
back under the `WKWebView` after the swipe is complete. The end result is that scroll pockets are
*removed* from the view hierarchy, since UIKit determines they aren't necessary, due to the fact
that the pocketed element rects don't intersect with the scroll view's frame right after the swipe
finishes, since the scroll view is offscreen.

As a temporary workaround, we forcibly invalidate the scroll pockets' cached geometries after a
navigation swipe finishes. This change should be reverted once the fix for the underlying issue
lands, in rdar://156271879.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _navigationGestureDidEnd]):
(-[WKWebView _forceScrollPocketsToRecomputeElementRegions]):
* Source/WebKit/UIProcess/ios/UIKitUtilities.h:

Canonical link: https://commits.webkit.org/297668@main
